libsawtooth/src/client/mod.rs

```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright 2018 Intel Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * ------------------------------------------------------------------------------
+ */
+
+mod error;
+
+use reqwest::{blocking::Client, Url};
+use serde::Deserialize;
+use std::collections::VecDeque;
+
+pub use error::SawtoothClientError;
+
+/// Client for Sawtooth Cli transactions
+/// Handles communication with the REST API
+pub struct RestApiSawtoothClient {
+    url: String,
+}
+
+trait SawtoothClient {
+    fn new(url: &str) -> Self;
+    fn list_batches(&self) -> Result<Box<dyn Iterator<Item = Batch>>, SawtoothClientError>;
+}
+
+impl SawtoothClient for RestApiSawtoothClient {
+    /// Create a new `RestApiSawtoothClient` with the given base `url`. The URL should be the bind
+    /// endpoint of the Sawtooth REST API
+    fn new(url: &str) -> RestApiSawtoothClient {
+        RestApiSawtoothClient { url: url.into() }
+    }
+
+    /// Lists all existing batches
+    fn list_batches(&self) -> Result<Box<dyn Iterator<Item = Batch>>, SawtoothClientError> {
+        let url = format!("{}/batches", &self.url);
+
+        Ok(Box::new(BatchIter::new(&url).unwrap()))
+    }
+}
+
+/// Iterator used for parsing and deserializing batches
+pub struct BatchIter {
+    next: String,
+    cache: VecDeque<Batch>,
+}
+
+impl BatchIter {
+    /// Create a new 'BatchIter' which will make a call to the REST API and load the initial
+    /// cache with the first page of batch entries
+    fn new(url: &str) -> Result<Self, SawtoothClientError> {
+        let url = Url::parse(url)
+            .map_err(|err| SawtoothClientError::new_with_source("invalid URL", err.into()))?;
+
+        let request = Client::new().get(url);
```

Comment:
This can be replaced with doing:
```rust
let mut new_iter = Self { next: Some(url.to_string()), cache: VecDec::with_capacity(0) }
new_iter.reload_cache()?;
Ok(new_iter)
```

Reply:
Fixed